File: android/app/src/main/java/com/snappet/mobile/feature/kilter/KilterGeneratorAssets.kt
package com.snappet.mobile.feature.kilter

import android.content.Context
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import java.io.File
import java.net.HttpURLConnection
import java.net.URL

/** Default host for the generator assets — the board-explorer's `climb-generator/` dir (same host the
 *  catalog downloads from; #42 user-hosted posture). One GET each for manifest/model/meta; nothing uploaded. */
const val KILTER_GENERATOR_HOST = "https://harshal2802.github.io/Snappet/climb-generator/"

/** The host's `manifest.json` — available generator models; the app installs the `default` one. */
@Serializable
data class KilterGeneratorManifest(val default: String, val models: List<Entry>) {
    @Serializable
    data class Entry(
        val id: String,
        val label: String,
        val model: String,
        val meta: String,
        val sizeBytes: Long? = null,
        val description: String? = null,
    )
}

/**
 * Lazy-downloads + caches the on-device generator's model + meta (~9 MB `model.q.onnx` + `meta.json`) into
 * app storage, so the binary stays slim and the model is updatable host-side. [ensureInstalled] is
 * idempotent. No ONNX here (Foundation/HTTP only). Mirrors iOS `KilterGeneratorAssets`.
 */
class KilterGeneratorAssets(context: Context, private val baseUrl: String = KILTER_GENERATOR_HOST) {
    private val dir = File(context.filesDir, "kilter-generator").apply { mkdirs() }
    val modelFile = File(dir, "model.onnx")
    val metaFile = File(dir, "meta.json")
    private val json = Json { ignoreUnknownKeys = true }

    val isInstalled: Boolean get() = modelFile.exists() && metaFile.exists()

    /** Ensure the model + meta are cached, downloading on first use; returns the decoded model. [progress] 0→1. */
    suspend fun ensureInstalled(progress: (Float) -> Unit = {}): KilterGeneratorModel = withContext(Dispatchers.IO) {
        if (isInstalled) {
            progress(1f)
            return@withContext KilterGeneratorModel.parse(metaFile.readText())
        }
        val manifest = fetchManifest()
        val entry = manifest.models.firstOrNull { it.id == manifest.default }
            ?: manifest.models.firstOrNull()
            ?: throw KilterGeneratorException("The generator manifest had no models.")
        download(entry.meta, metaFile) { progress(it * 0.05f) }
        download(entry.model, modelFile) { progress(0.05f + it * 0.95f) }
        progress(1f)
        KilterGeneratorModel.parse(metaFile.readText())
    }

    fun remove() {
        modelFile.delete()
        metaFile.delete()
    }

    private fun fetchManifest(): KilterGeneratorManifest {
        val conn = URL(baseUrl + "manifest.json").openConnection() as HttpURLConnection
        conn.connectTimeout = 30_000; conn.readTimeout = 60_000
        try {
            if (conn.responseCode !in 200..299) {
                throw KilterGeneratorException("Couldn't reach the generator host (HTTP ${conn.responseCode}).")
            }
            return json.decodeFromString(KilterGeneratorManifest.serializer(), conn.inputStream.bufferedReader().readText())
        } finally {
            conn.disconnect()
        }
    }

    /** Stream a file to [dest] with progress (atomic-ish swap via a `.part` temp). */
    private fun download(name: String, dest: File, progress: (Float) -> Unit) {
        val conn = URL(baseUrl + name).openConnection() as HttpURLConnection
        conn.connectTimeout = 30_000; conn.readTimeout = 120_000
        try {
            if (conn.responseCode !in 200..299) {
                throw KilterGeneratorException("Couldn't reach the generator host (HTTP ${conn.responseCode}).")
            }
            val total = conn.contentLengthLong
            val tmp = File(dest.path + ".part")
            conn.inputStream.use { input ->
                tmp.outputStream().use { output ->
                    val buf = ByteArray(1 shl 16)
                    var received = 0L
                    var n: Int
                    while (input.read(buf).also { n = it } >= 0) {
                        output.write(buf, 0, n)
                        received += n
                        if (total > 0) progress((received.toFloat() / total).coerceIn(0f, 1f))
                    }
                }
            }
            if (tmp.length() == 0L) throw KilterGeneratorException("The download was empty. Try again.")
            dest.delete()
            if (!tmp.renameTo(dest)) throw KilterGeneratorException("Couldn't save the model on this device.")
        } finally {
            conn.disconnect()
        }
    }
}

File: ios/App/Snappet/Features/Kilter/KilterGeneratorAssets.swift
import Foundation

/// Default host for the generator assets — the board-explorer's `climb-generator/` directory (same
/// GitHub Pages site the catalog downloads from). One GET each for the manifest, model, and meta; nothing
/// is uploaded. See `kilterDefaultCatalogHost` for the catalog equivalent + the legal posture (#42).
let kilterGeneratorHost = "https://harshal2802.github.io/Snappet/climb-generator/"

/// The host's `manifest.json` — the available generator models. The app installs the `default` one.
struct KilterGeneratorManifest: Codable, Sendable {
    let `default`: String
    let models: [Entry]

    struct Entry: Codable, Sendable, Identifiable {
        let id: String
        let label: String
        let model: String     // filename of the .onnx
        let meta: String      // filename of the meta.json
        let sizeBytes: Int64?
        let description: String?
    }
}

enum KilterGeneratorAssetError: LocalizedError {
    case http(Int)
    case emptyManifest
    case emptyDownload

    var errorDescription: String? {
        switch self {
        case .http(let code): return "Couldn't reach the generator host (HTTP \(code)). Check your connection."
        case .emptyManifest: return "The generator manifest had no models."
        case .emptyDownload: return "The model download was empty. Try again."
        }
    }
}

/// Lazy-downloads + caches the on-device generator's model + meta (the ~9 MB `model.q.onnx` and
/// `meta.json`) into Application Support, so the app binary stays slim and the model can be updated host-
/// side without an app release. Mirrors the catalog's `HostedCatalogClient` download posture (streamed
/// GETs with progress, ephemeral session). `ensureInstalled` is idempotent — once cached it returns the
/// decoded `KilterGeneratorModel` without touching the network. Foundation only (no ONNX).
final class KilterGeneratorAssets: Sendable {
    static let shared = KilterGeneratorAssets()

    private let baseURL: String
    private let session: URLSession
    private let dir: URL
    /// Records which model id is cached, so a manifest default-change re-downloads.
    private let stampURL: URL

    init(baseURL: String = kilterGeneratorHost) {
        self.baseURL = baseURL.hasSuffix("/") ? baseURL : baseURL + "/"
        let support = (try? FileManager.default.url(for: .applicationSupportDirectory,
                                                    in: .userDomainMask, appropriateFor: nil, create: true))
            ?? FileManager.default.temporaryDirectory
        self.dir = support.appendingPathComponent("KilterGenerator", isDirectory: true)
        try? FileManager.default.createDirectory(at: dir, withIntermediateDirectories: true)
        self.stampURL = dir.appendingPathComponent("installed.json")

        let config = URLSessionConfiguration.ephemeral
        config.timeoutIntervalForRequest = 120
        config.waitsForConnectivity = true
        self.session = URLSession(configuration: config)
    }

    var modelURL: URL { dir.appendingPathComponent("model.onnx") }
    var metaURL: URL { dir.appendingPathComponent("meta.json") }

    /// True when both assets are already on disk (so the Generate tab can skip the download UI).
    var isInstalled: Bool {
        FileManager.default.fileExists(atPath: modelURL.path)
            && FileManager.default.fileExists(atPath: metaURL.path)
    }

    /// Ensure the model + meta are cached, downloading them on first use, and return the decoded model.
    /// Idempotent; `progress` runs 0→1 over the (meta + model) download (skipped when already installed).
    func ensureInstalled(progress: @escaping @Sendable (Double) -> Void = { _ in }) async throws -> KilterGeneratorModel {
        if isInstalled, let model = try? KilterGeneratorModel.load(metaURL: metaURL) {
            progress(1)
            return model
        }
        let manifest = try await fetchManifest()
        guard let entry = manifest.models.first(where: { $0.id == manifest.default })
                ?? manifest.models.first else {
            throw KilterGeneratorAssetError.emptyManifest
        }
        // Meta is tiny (~150 KB); model is ~9 MB — weight progress accordingly.
        try await download(entry.meta, to: metaURL) { progress($0 * 0.05) }
        try await download(entry.model, to: modelURL) { progress(0.05 + $0 * 0.95) }
        writeStamp(modelId: entry.id)
        progress(1)
        return try KilterGeneratorModel.load(metaURL: metaURL)
    }

    /// Delete the cached assets (e.g. from Settings, to reclaim space).
    func remove() {
        try? FileManager.default.removeItem(at: modelURL)
        try? FileManager.default.removeItem(at: metaURL)
        try? FileManager.default.removeItem(at: stampURL)
    }

    // MARK: - Networking

    private func fetchManifest() async throws -> KilterGeneratorManifest {
        guard let url = URL(string: baseURL + "manifest.json") else { throw KilterGeneratorAssetError.emptyManifest }
        let (data, resp) = try await session.data(from: url)
        if let http = resp as? HTTPURLResponse, !(200..<300).contains(http.statusCode) {
            throw KilterGeneratorAssetError.http(http.statusCode)
        }
        return try JSONDecoder().decode(KilterGeneratorManifest.self, from: data)
    }

    /// Stream a file to `dest` with progress, mirroring `HostedCatalogClient.download`.
    private func download(_ filename: String, to dest: URL,
                          progress: @escaping @Sendable (Double) -> Void) async throws {
        let url = URL(string: baseURL + filename) ?? URL(fileURLWithPath: filename)
        let (bytes, response) = try await session.bytes(from: url)
        guard let http = response as? HTTPURLResponse else { throw KilterGeneratorAssetError.http(0) }
        guard (200..<300).contains(http.statusCode) else { throw KilterGeneratorAssetError.http(http.statusCode) }

        let tmp = dest.appendingPathExtension("part")
        FileManager.default.createFile(atPath: tmp.path, contents: nil)
        let handle = try FileHandle(forWritingTo: tmp)
        defer { try? handle.close() }

        let total = http.expectedContentLength
        var received: Int64 = 0
        var buffer = Data()
        buffer.reserveCapacity(1 << 16)
        for try await byte in bytes {
            buffer.append(byte)
            if buffer.count >= (1 << 16) {
                // Throwing write, not the legacy `write(_:)` — disk-full mid-download must surface as
                // a thrown error, not an uncatchable NSFileHandleOperationException.
                try handle.write(contentsOf: buffer); received += Int64(buffer.count); buffer.removeAll(keepingCapacity: true)
                if total > 0 { progress(min(1, Double(received) / Double(total))) }
            }
        }
        if !buffer.isEmpty { try handle.write(contentsOf: buffer); received += Int64(buffer.count) }
        try? handle.close()
        guard received > 0 else { throw KilterGeneratorAssetError.emptyDownload }
        // Atomic-ish swap so a half-written file is never treated as installed.
        try? FileManager.default.removeItem(at: dest)
        try FileManager.default.moveItem(at: tmp, to: dest)
        progress(1)
    }

    private func writeStamp(modelId: String) {
        let json = try? JSONSerialization.data(withJSONObject: ["modelId": modelId])
        try? json?.write(to: stampURL)
    }
}
